Change objectsAtCoord in MaplyBaseViewController to search for markers, labels and vector objects

akovalov opened the pull request:
This improvement is for https://github.com/mousebird-consulting-inc/WhirlyGlobe/issues/1489
- Exposed `selectMultipleLabelsAndMarkersForScreenPoint` function in `MaplyBaseInteractionLayer` to be able to call it from outside.
- Moved same code that converts `SelectionManager::SelectedObject` to `MaplySelectedObject` from `WGInteractionLayer.m` and `MaplyInteractionLayer.mm` to avoid code duplication. Exposed `convertSelectedObjects` and `convertSelectedVecObjects` in `MaplyBaseInteractionLayer` for this matter to use in subclasses.
- Changed `objectsAtCoord` function in `MaplyBaseViewController` to return markers, labels and vector objects at a coordinate in one array. It's one common function implementation for both Maply and WhirlyGlobe. 
- Added long press gesture to MarkersTestCase to test the `objectsAtCoord` function.

TimSylvester-mousebird commented:
If `objectsAtCoord` is going to search both vector objects and screen objects, that should be optional.  There should remain a way to search each set individually.

akovalov commented:
@TimSylvester-mousebird thank you for your comment, it makes sense. 

Ok, reverted `objectsAtCoord` to return only vector objects. But it didn't work because `return [renderControl->interactLayer findVectorsInPoint:Point2f(coord.x,coord.y)];` under the hood passed `nil` in `inView:(MaplyBaseViewController *)vc` inside `MaplyBaseInteractionLayer.mm` which does not make sense because of the following checking where `vc` should not be nil. 

```
- (NSArray *)findVectorsInPoint:(Point2f)pt inView:(MaplyBaseViewController *)vc multi:(bool)multi
{
    if (!layerThread || !vc || !vc->renderControl)
        return nil;
...
```

So call `findVectorsInPoint:Point2f(coord.x,coord.y)]` without providing a view should not be used, because will always return nil. There are such calls without a view, however `WGInteractionLayer` has a correct call with view in `userDidTapLayerThread` method. So I've removed variant of method without a view and added passing view where needed. Let me know please if this is a wrong idea. 

Added separate `- (NSArray * _Nullable)labelsAndMarkersAtCoord:(MaplyCoordinate)coord` and added explicit `vector objects` in `Return all the selectable vector objects at the given location.` doc comment for `objectsAtCoord`.

TimSylvester-mousebird commented on ios/library/WhirlyGlobe-MaplyComponent/include/private/MaplyBaseInteractionLayer_private.h:
```diff
@@ -205,10 +205,13 @@
+// Find MaplySelectableObjects at a screen point
 - (NSObject*__nullable)selectLabelsAndMarkerForScreenPoint:(CGPoint)screenPoint;
+- (NSMutableArray*__nullable)selectMultipleLabelsAndMarkersForScreenPoint:(CGPoint)screenPoint;
+- (NSMutableArray*__nullable)convertSelectedObjects:(std::vector<WhirlyKit::SelectionManager::SelectedObject>)selectedObjs;
```
The collection should be passed by constant reference (`const .. &`) here to avoid making an unnecessary copy.

akovalov replied:
Thank you @TimSylvester-mousebird for the review. I haven't a lot c++ experience, so please let me know if fix is not correct.

TimSylvester-mousebird commented on ios/library/WhirlyGlobe-MaplyComponent/src/control/MaplyBaseInteractionLayer.mm:
```diff
@@ -3664,7 +3659,58 @@ - (NSObject *)getSelectableObject:(WhirlyKit::SimpleIdentity)objId
+        // Work through the objects the manager found, creating entries for each
+        for (unsigned int ii=0;ii<selectedObjs.size();ii++)
+        {
+            SelectionManager::SelectedObject &theSelObj = selectedObjs[ii];
```
this will also need to be a `const` reference

TimSylvester-mousebird commented on ios/library/WhirlyGlobe-MaplyComponent/src/control/MaplyBaseInteractionLayer.mm:
```diff
@@ -3664,7 +3659,58 @@ - (NSObject *)getSelectableObject:(WhirlyKit::SimpleIdentity)objId
+- (NSMutableArray*)convertSelectedObjects:(std::vector<SelectionManager::SelectedObject>)selectedObjs
+{
+    NSMutableArray *retSelectArr = [NSMutableArray array];
+    if (!selectedObjs.empty())
```
This isn't really necessary, as a for-loop over an empty collection is no slower.

TimSylvester-mousebird reviewed (changes requested):
So `objectsAtCoord` wasn't working anyway.  At least we don't have to worry about breaking anything using it. :grin:

I'm not seeing tap feedback on the marker test case sometimes, but that looks to be a bug in the testcase itself, and the hit-testing is working.

I've just flagged one minor C++ issue, otherwise it looks great, thanks for the PR!

TimSylvester-mousebird reviewed (approved):
Looks great, thanks again!
